### ISSUE TITLE
Sources without some fields

### ISSUE THREAD

**Issue by smimram** [savonet/liquidsoap#3203]
The syntax
```
s = (single("test.png"):source(audio=none))
```
for specifying that there is no audio is not working anymore. How do we do this now?

**Comment by toots** [comment]
You can specify content explicitely:
```liquidsoap
s = (single("test.png"):source(video=canvas))
```
There is a `never` type now, used for optional methods but it's not wired as a type annotation. We could if you think it could be useful.

**Comment by smimram** [comment]
Yes it would be useful for instance in order to prevent audio channel from being present without having to specify precisely all other channels (video can be pretty complicated).

**Comment by toots** [comment]
Okay. The more general question from a type annotation standpoint is: how can we allow the description of an open type? Something like `'a.{foo:int}` ?

**Comment by Moonbase59** [comment]
Let me throw a word in: I have videos containing more than 10 audio and 30-something subtitle tracks. How could we address a certain audio/subtitle track? (They are usually identified by 2- or 3-letter ISO language codes.)

Just a thought, for future expansion, so it doesn’t get forgotten. ;-)

**Comment by toots** [comment]
> Let me throw a word in: I have videos containing two video, more than 10 audio and 30-something subtitle tracks. How could we address a certain audio/subtitle track? (They are usually identified by 2- or 3-letter ISO language codes.)
> 
> Just a thought, for future expansion, so it doesn’t get forgotten. ;-)
> 
> Eventually, we’ll probably need
> 
> * some _selector_ that can select one or more track types (video, audio, subtitle) based on its language code
> * some _deselector_ for track types (no audio, no video, no subtitles, like the old _drop_xxx_ functions)
> 
> Ideally, things like this should be possible (far future…):
> 
> * build a source from
>   
>   * the video track
>   * the English, German and French audio track (if existent)
>   * English, German, French subtitles
>     
>     * which should also be selectable by flags (default, hearing impaired, visually impaired, commentary, etc.)
> * and mux a radio stream into it as an additional audio track
> * set a "default" set of all types (including "none") for initial playout
> * video player clients can then select a track combination according to their/the user’s preference
> 
> Just a simple example. Don’t know if it’s feasible, since it would require realtime remuxing for streaming. A preferable input format could be MKV (Matroska), outputting to either H.264/H.265 MPEG-TS or HLS for streaming.

The documentation for muxing/demuxing is here: https://www.liquidsoap.info/doc-dev/multitrack.html#tracks-demuxing-and-muxing

**Comment by toots** [comment]
Also, there's no way to select based on track metadata (language, codec, channels etc). I plan on adding that in the future.

**Comment by Moonbase59** [comment]
No problem, as I said: »Just a thought, for future expansion« :-)

**Comment by smimram** [comment]
I think that instead of `'a.{foo:int}` the following notation is much better: `{foo: int, ...}`.

Agreed it's less precise (we cannot reuse the `'a`) but it should be enough for most uses for now!
